File: javascript/kindergarten-garden/kindergarten-garden.js
const plants = {
  C: 'clover',
  G: 'grass',
  R: 'radishes',
  V: 'violets'
};

const defaultStudents = [
  'Alice',
  'Bob',
  'Charlie',
  'David',
  'Eve',
  'Fred',
  'Ginny',
  'Harriet',
  'Ileana',
  'Joseph',
  'Kincaid',
  'Larry'
];

const parse = diagram => diagram.split('\n').map(row => [...row].map(letter => plants[letter]));

const getPlants = (parsedDiagram, index) => {
  const studentPlants = [];
  const position = index * 2;

  studentPlants.push(parsedDiagram[0][position]);
  studentPlants.push(parsedDiagram[0][position + 1]);
  studentPlants.push(parsedDiagram[1][position]);
  studentPlants.push(parsedDiagram[1][position + 1]);

  return studentPlants;
};

class Garden {
  constructor(diagram, students) {
    this.students = students || defaultStudents;
    this.students.sort();

    this.students.forEach((student, index) => {
      this[student.toLowerCase()] = getPlants(parse(diagram), index);
    });
  }
}

if (module) { module.exports = Garden; }

File: ruby/kindergarten-garden/kindergarten_garden.rb
PLANTS = { 'C' => :clover, 'G' => :grass, 'R' => :radishes, 'V' => :violets }.freeze

STUDENTS = %w[Alice Bob Charlie David Eve Fred Ginny Harriet Ileana Joseph Kincaid Larry].freeze

class Garden
  def initialize(diagram, students = nil)
    @rows = diagram.split("\n").map { |row| row.chars.map { |char| PLANTS[char] } }
    @students = students ? students.sort : STUDENTS
  end

  def plants_for(student)
    plants = []
    position = @students.index(student) * 2
    @rows.each do |row|
      plants << row[position]
      plants << row[position + 1]
    end
    plants
  end

  def method_missing(method_name, *args, &block)
    possible_student = method_name.id2name.capitalize
    if @students.include?(possible_student)
      plants_for(possible_student)
    else
      super
    end
  end

  def respond_to_missing?(method_name, *args)
    @students.include?(method_name.id2name.capitalize) || super
  end
end
